File: NNThemeKit/ThemeModeObserver.m
//
//  ThemeModeObserver.m
//  NNThemeKit
//
//  Created by NeroXie on 2025/11/6.
//

#import "ThemeModeObserver.h"
#import "ThemeIMP.h"
#import "ThemeModel.h"
#import "UIView+Theme.h"

NSString * const ThemeModeChangedNotification = @"theme.modeChanged";

static NSString * const kThemeModeKey = @"ThemeMode";

API_AVAILABLE(ios(13.0))
@interface ThemeObserveWindow : UIWindow

@end

@interface ThemeModeObserver()

@property (nonatomic, strong) NSHashTable<UIView *> *viewHashTable;
@property (nonatomic, assign) ThemeMode currentMode;
@property (nonatomic, assign) UIUserInterfaceStyle currentUserInterfaceStyle;
@property (nonatomic, strong, nullable) UIWindow *themeObserveWindow;

@end

@implementation ThemeModeObserver

+ (instancetype)sharedInstance {
    static id _sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _sharedInstance = [self new];
    });
    
    return _sharedInstance;
}

- (instancetype)init {
    if (self = [super init]) {
        _viewHashTable = [NSHashTable hashTableWithOptions:NSHashTableWeakMemory];
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(_updateNewWindowUserInterfaceStyle:)
                                                     name:UIWindowDidBecomeVisibleNotification
                                                   object:nil];
    }
    return self;
}

- (void)setup {
    ThemeMode mode = ThemeModeLight;
    NSNumber *numberValue = [[NSUserDefaults standardUserDefaults] valueForKey:kThemeModeKey];
    if (numberValue) mode = [numberValue integerValue];
    [self changeThemeMode:mode];
}

- (BOOL)isDarkMode {
    switch (self.currentMode) {
        case ThemeModeLight:
            return NO;
        case ThemeModeDark:
            return YES;
        case ThemeModeSystem:
            return self.currentUserInterfaceStyle == UIUserInterfaceStyleDark;
    }
}

//- (UIStatusBarStyle)preferredStatusBarStyle {
//    switch (self.currentMode) {
//        case ThemeModeLight:
//            return UIStatusBarStyleDarkContent;
//        case ThemeModeDark:
//            return UIStatusBarStyleLightContent;
//        case ThemeModeSystem:
//        default:
//            return UIStatusBarStyleDefault;
//    }
//}

- (void)changeThemeMode:(ThemeMode)themeMode {
    NSLog(@"%s", __func__);
    // 不支持深色模式时只能支持为 Light Mode
    if (!ThemeIMP.sharedInstance.data.supportDarkMode && themeMode != ThemeModeLight) {
        themeMode = ThemeModeLight;
    }
    
    _currentMode = themeMode;
    
    [[NSUserDefaults standardUserDefaults] setInteger:themeMode forKey:kThemeModeKey];
    [[NSUserDefaults standardUserDefaults] synchronize];
    
    switch (themeMode) {
        case ThemeModeLight:
            [self _changeAllWindowsUserInterfaceStyle:UIUserInterfaceStyleLight];
            [self _changeCurrentUserInterfaceStyle:UIUserInterfaceStyleLight];
            break;
        case ThemeModeDark:
            [self _changeAllWindowsUserInterfaceStyle:UIUserInterfaceStyleDark];
            [self _changeCurrentUserInterfaceStyle:UIUserInterfaceStyleDark];
            break;
        case ThemeModeSystem:
            [self _changeAllWindowsUserInterfaceStyle:UIUserInterfaceStyleUnspecified];
            [self _changeCurrentUserInterfaceStyle:UITraitCollection.currentTraitCollection.userInterfaceStyle];
            break;
    }
}

- (void)cacheThemeUIActionsWithView:(UIView *)view {
    if (![self.viewHashTable containsObject:view]) {
        [self.viewHashTable addObject:view];
    }
}


#pragma mark - Private Method

- (void)_changeAllWindowsUserInterfaceStyle:(UIUserInterfaceStyle)userInterfaceStyle {
    NSLog(@"%s", __func__);
    for (UIWindow *window in UIApplication.sharedApplication.windows) {
        if (self.themeObserveWindow && window == self.themeObserveWindow) continue;
        window.overrideUserInterfaceStyle = userInterfaceStyle;
    }
}

- (void)_changeCurrentUserInterfaceStyle:(UIUserInterfaceStyle)style {
    NSLog(@"%s", __func__);
    
    _currentUserInterfaceStyle = style;
    
    [NSNotificationCenter.defaultCenter postNotificationName:ThemeModeChangedNotification object:nil];
    
    for (UIView *view in self.viewHashTable.allObjects) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wundeclared-selector"
        if (view && [view respondsToSelector:@selector(_theme_reloadUIActions)]) {
            [view performSelector:@selector(_theme_reloadUIActions)];
        }
#pragma clang diagnostic pop
    }
    
    [[NSNotificationCenter defaultCenter] postNotificationName:ThemeChangedSuccessNotification object:@(style)];
}

- (void)_updateNewWindowUserInterfaceStyle:(NSNotification *)notification {
    UIWindow *window = notification.object;
    if (![window isKindOfClass:[UIWindow class]]) return;
    
    if (window.traitCollection.userInterfaceStyle != self.currentUserInterfaceStyle) {
        window.overrideUserInterfaceStyle = self.currentUserInterfaceStyle;
    }
    
    if (UIApplication.sharedApplication.delegate.window == window) {
        self.themeObserveWindow = [[ThemeObserveWindow alloc] initWithFrame:CGRectZero];
        self.themeObserveWindow.windowLevel = UIWindowLevelNormal + 1;
        self.themeObserveWindow.overrideUserInterfaceStyle = UIUserInterfaceStyleUnspecified;
    }
}

@end

@implementation ThemeObserveWindow

- (void)traitCollectionDidChange:(UITraitCollection *)previousTraitCollection {
    if ([self.traitCollection hasDifferentColorAppearanceComparedToTraitCollection:previousTraitCollection]) {
        ThemeModeObserver *instance = ThemeModeObserver.sharedInstance;
        UIUserInterfaceStyle userInterfaceStyle = self.traitCollection.userInterfaceStyle;
        if (instance.currentMode == ThemeModeSystem && userInterfaceStyle != instance.currentUserInterfaceStyle) {
            [instance _changeCurrentUserInterfaceStyle:userInterfaceStyle];
        }
    }
}

@end

File: NNThemeKit/Classes/ThemeModeObserver.swift
//
//  ThemeMode.swift
//  Example_ThemeManager
//
//  Created by NeroXie on 2021/6/22.
//

import Foundation

@objc public enum ThemeMode: Int, Equatable {
    
    case system = 0
    
    case light = 1
    
    case dark = 2
}

@available(iOS 13.0, *)
class ThemeModeObserver {
    
    static let shared = ThemeModeObserver()
    
    private var viewHashTable = NSHashTable<UIView>(options: .weakMemory)
    
    private(set) var currentMode: ThemeMode = .system
    
    private(set) var currentUserInterfaceStyle: UIUserInterfaceStyle = .unspecified
    
    private var themeObserveWindow: UIWindow?
    
    var isDarkMode: Bool {
        switch currentMode {
        case .light: return false
        case .dark: return true
        case .system: return currentUserInterfaceStyle == .dark
        }
    }
    
    var preferredStatusBarStyle: UIStatusBarStyle {
        switch currentMode {
        case .light: return .darkContent
        case .dark: return .lightContent
        case .system: return .default
        }
    }
    
    private init() {}
    
    func setup() {
        var themeMode = currentMode
        let rawValue = UserDefaults.standard.value(forKey: .themeModeKey)
        if let _rawValue = rawValue as? Int, let cacheMode = ThemeMode(rawValue: _rawValue) {
            themeMode = cacheMode
        }
        
        if ThemeIMP.shared.data.supportDarkMode {
            let selector = #selector(changeWindowsUserInterfaceStyle(with:))
            let name = UIWindow.didBecomeVisibleNotification
            NotificationCenter.default.addObserver(self, selector: selector, name: name, object: nil)
        } else {
            themeMode = .light
        }
        
        changeThemeMode(themeMode)
    }
    
    func changeThemeMode(_ themeMode: ThemeMode) {
        currentMode = themeMode
        UserDefaults.standard.setValue(themeMode.rawValue, forKey: .themeModeKey)
        UserDefaults.standard.synchronize()
        switch themeMode {
        case .light:
            updateAllWindowUserInterfaceStyle(.light)
            updateCurrentUserInterfaceStyle(.light)
        case .dark:
            updateAllWindowUserInterfaceStyle(.dark)
            updateCurrentUserInterfaceStyle(.dark)
        case .system:
            updateAllWindowUserInterfaceStyle(.unspecified)
            updateCurrentUserInterfaceStyle(UITraitCollection.current.userInterfaceStyle)
        }
    }
    
    func cacheThemeUIActions(with view: UIView) {
        guard !viewHashTable.contains(view) else { return }
        
        viewHashTable.add(view)
    }
    
    // MARK: - Private Method
    
    private func updateCurrentUserInterfaceStyle(_ style: UIUserInterfaceStyle) {
        currentUserInterfaceStyle = style
        ThemeIMP.shared.reloadThemeColors(with: isDarkMode)
        reloadThemeUIActions()
        NotificationCenter.default.post(name: Theme.changedThemeSuccessNotification, object: style)
    }
    
    private func updateAllWindowUserInterfaceStyle(_ userInterfaceStyle: UIUserInterfaceStyle) {
        for window in UIApplication.shared.windows {
            if let theWindow = themeObserveWindow, theWindow == window {
                continue
            }
            
            window.overrideUserInterfaceStyle = userInterfaceStyle
        }
    }
    
    fileprivate func traitCollectionDidChange(with style: UIUserInterfaceStyle) {
        guard currentMode == ThemeMode.system, style != currentUserInterfaceStyle else {
            return
        }
        
        updateCurrentUserInterfaceStyle(style)
    }
    
    @objc private func changeWindowsUserInterfaceStyle(with notification: NSNotification) {
        guard let window = notification.object as? UIWindow else { return }
        if window.traitCollection.userInterfaceStyle != currentUserInterfaceStyle {
            window.overrideUserInterfaceStyle = currentUserInterfaceStyle
        }
        
        if let delegateWindow = UIApplication.shared.delegate?.window, delegateWindow == window {
            themeObserveWindow = ThemeObserveWindow(frame: .zero)
            themeObserveWindow?.windowLevel = .normal + 1
            themeObserveWindow?.overrideUserInterfaceStyle = .unspecified
        }
    }
    
    private func reloadThemeUIActions() {
        print(#function)
        //        let start = CFAbsoluteTimeGetCurrent()
        viewHashTable.allObjects.forEach { $0.theme.reloadThemeUIActions() }
        //        let end = CFAbsoluteTimeGetCurrent()
        //        print("Time of reloading UI action：\((end-start) * 1000) 毫秒")
        //        if reloadAllThemeAction {
        //            viewHashTable.allObjects.forEach { $0.t.reloadThemeUIActions() }
        //        } else {
        //            viewHashTable.allObjects
        //            // .filter { $0.t.isDisplayedInScreen }
        //                .forEach { $0.t.reloadThemeUIActions() }
        //        }
    }
}

fileprivate extension String {
    
    @available(iOS 13.0, *)
    static var themeModeKey: String { "ThemeMode" }
}

@available(iOS 13.0, *)
fileprivate class ThemeObserveWindow: UIWindow {
    
    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
        guard traitCollection.hasDifferentColorAppearance(comparedTo: previousTraitCollection) else {
            return
        }
        
        ThemeModeObserver.shared.traitCollectionDidChange(with: traitCollection.userInterfaceStyle)
    }
}
